src/ThaiRomanizationSharp.Thai2Rom/NeuralNet/Attn.cs

```diff
@@ -0,0 +1,64 @@
+﻿using System;
+using TorchSharp.Modules;
+using static TorchSharp.torch;
+using F = TorchSharp.torch.nn.functional;
+
+namespace ThaiRomanizationSharp.Thai2Rom.NeuralNet
+{
+    internal sealed class Attn : nn.CustomModule
+    {
+        private readonly string method;
+        private readonly long hiddenSize;
+        private readonly Linear attn;
+
+        public Attn(string method, long hiddenSize)
+            :base("attn")
+        {
+            this.method = method;
+            this.hiddenSize = hiddenSize;
+
+            if(this.method == "general")
+            {
+                this.attn = nn.Linear(this.hiddenSize, hiddenSize);
+            }
+            else// if(this.method == "concat")
+            {
+                throw new NotImplementedException("not used in original python program");
+            }
+
+            RegisterComponents();
+        }
+
+        internal Tensor Forward(Tensor hidden, Tensor encoderOutputs, Tensor mask)
+        {
+            // Calculate energies for each encoder output
+            Tensor attnEnergies;
+            if(this.method == "dot")
+            {
+                throw new NotImplementedException("not used in original python program");
+            }
+            else if (this.method == "general")
+            {
+                attnEnergies = this.attn.forward(
+                    encoderOutputs.view(-1, encoderOutputs.size(-1))
+                ); // (batch_size * sequence_len,  hidden_size)
+                attnEnergies = attnEnergies
+                    .view(encoderOutputs.size())
+                    .bmm(
+                        hidden.transpose(1, 2))
+                    .squeeze(
+                        2
+                    );
+            }
+            else// if (this.method == "concat")
```

Comment:
Not sure you have intension to keep this statement `if (this.method == "concat")
Write`?

Reply:
I do intend to keep it; it's part of the python code that I didn't need. I'll update this to be more obvious, thanks!